### ISSUE TITLE
dvc status: wrong etag used when computing status for imported stage

### ISSUE THREAD

**Issue by kwon-young** [iterative/dvc#5998]
# Bug Report

## Description

After importing data from a remote url using `dvc import-url https://raw.githubusercontent.com/musescore/MuseScore/master/fonts/leland/metadata.json`, a dvc file `metadata.json.dvc` is created together with `myremotefile`.
In this dvc file (ex shown below), the etag of the file is saved:  

```
md5: 6b4495b523ff272b9ca3de0a052f90a9
frozen: true
deps:
- etag: W/"f3810bf243557a80191cc0b03e61e0b4278179cf79805b219e8f4f2b3e9cb252"
  size: 9789
  path: https://raw.githubusercontent.com/musescore/MuseScore/master/fonts/leland/metadata.json
outs:
- md5: c6d0a7e1a1828d825ff75ccbb67be72d
  size: 116976
  path: metadata.json

```

However, doing a `dvc status` shows that the file checksum has changed, and I have also added a strategic print statement in [dvc/utils/__init__.py](https://github.com/iterative/dvc/blob/master/dvc/utils/__init__.py#L100) to show exactly what data is used to compute the md5sum:

```
$ dvc status data/Leland/metadata.json -v
2021-05-12 14:17:10,723 DEBUG: Check for update is enabled.
2021-05-12 14:17:10,745 DEBUG: Checking if stage 'data/Leland/metadata.json' is in 'dvc.yaml'
2021-05-12 14:17:11,097 WARNING: stage: 'data/Leland/metadata.json.dvc' is frozen. Its dependencies are not going to be shown in the status output.
{'deps': [{'etag': 'w/"f3810bf243557a80191cc0b03e61e0b4278179cf79805b219e8f4f2b3e9cb252"',
           'path': 'https://raw.githubusercontent.com/musescore/MuseScore/master/fonts/leland/metadata.json'}],
 'outs': [{'md5': 'c6d0a7e1a1828d825ff75ccbb67be72d', 'path': 'metadata.json'}]}
2021-05-12 14:17:11,115 DEBUG: Computed stage: 'data/Leland/metadata.json.dvc' md5: 'fda525c81393b77ce3a41813bb93dea7'
2021-05-12 14:17:11,115 DEBUG: 'md5' of stage: 'data/Leland/metadata.json.dvc' changed.
data/Leland/metadata.json.dvc:
        changed checksum
2021-05-12 14:17:11,129 DEBUG: Analytics is disabled.
```

Already here, we can see a difference in how the etag is represented with

* `W/"f3810bf243557a80191cc0b03e61e0b4278179cf79805b219e8f4f2b3e9cb252"` in the dvc file
* `w/"f3810bf243557a80191cc0b03e61e0b4278179cf79805b219e8f4f2b3e9cb252"` for computing the md5sum of the dvc file.

Since `dvc status` says the md5sum of the dvc file changed, I tried to use `dvc commit` to update the md5sum of the file:

```
$ dvc commit data/Leland/metadata.json -v
2021-05-12 14:23:07,226 DEBUG: Check for update is enabled.
2021-05-12 14:23:07,255 DEBUG: Checking if stage 'data/Leland/metadata.json' is in 'dvc.yaml'
{'deps': [{'etag': 'w/"f3810bf243557a80191cc0b03e61e0b4278179cf79805b219e8f4f2b3e9cb252"',
           'path': 'https://raw.githubusercontent.com/musescore/MuseScore/master/fonts/leland/metadata.json'}],
 'outs': [{'md5': 'c6d0a7e1a1828d825ff75ccbb67be72d', 'path': 'metadata.json'}]}
2021-05-12 14:23:07,947 DEBUG: Computed stage: 'data/Leland/metadata.json.dvc' md5: 'fda525c81393b77ce3a41813bb93dea7'
2021-05-12 14:23:07,947 DEBUG: 'md5' of stage: 'data/Leland/metadata.json.dvc' changed.
dependencies ['https://raw.githubusercontent.com/musescore/MuseScore/master/fonts/leland/metadata.json'] of stage: 'data/Leland/metadata.json.dvc' changed. Are you sure you want to commit it? [y/n] y
2021-05-12 14:23:11,002 DEBUG: {}
2021-05-12 14:23:11,002 DEBUG: Output 'data/Leland/metadata.json' didn't change. Skipping saving.
{'deps': [{'etag': 'W/"f3810bf243557a80191cc0b03e61e0b4278179cf79805b219e8f4f2b3e9cb252"',
           'path': 'https://raw.githubusercontent.com/musescore/MuseScore/master/fonts/leland/metadata.json'}],
 'outs': [{'md5': 'c6d0a7e1a1828d825ff75ccbb67be72d', 'path': 'metadata.json'}]}
2021-05-12 14:23:11,002 DEBUG: Computed stage: 'data/Leland/metadata.json.dvc' md5: '6b4495b523ff272b9ca3de0a052f90a9'
2021-05-12 14:23:11,011 DEBUG: Checking out 'data/Leland/metadata.json' with cache 'object md5: c6d0a7e1a1828d825ff75ccbb67be72d'.
2021-05-12 14:23:11,012 DEBUG: Cache type 'reflink' is not supported: reflink is not supported
2021-05-12 14:23:11,014 DEBUG: Created 'copy': .dvc/cache/.cache_type_test_file -> data/Leland/.dGWVBvJHddU7r9xpMhMccD
2021-05-12 14:23:11,015 DEBUG: Removing 'data/Leland/.dGWVBvJHddU7r9xpMhMccD'
2021-05-12 14:23:11,018 DEBUG: Removing '.dvc/cache/.cache_type_test_file'
2021-05-12 14:23:11,018 DEBUG: Skipping copying for 'data/Leland/metadata.json', since it is not a symlink or a hardlink.
2021-05-12 14:23:11,018 DEBUG: state save (1100372401515, 1620814163718877952, 116976) c6d0a7e1a1828d825ff75ccbb67be72d
2021-05-12 14:23:11,042 DEBUG: Saving information to 'data/Leland/metadata.json.dvc'.
2021-05-12 14:23:11,064 DEBUG: Analytics is disabled.
```

Unfortunately, this did not work because dvc computes the new md5sum of the dvc file by doing a new request and using the headers with the wrongly formatted etag (with the uppercase W like this `W/"f3810bf243557a80191cc0b03e61e0b4278179cf79805b219e8f4f2b3e9cb252"`).

I did a quickfix by changing [this line](https://github.com/iterative/dvc/blob/master/dvc/fs/http.py#L151) to:

```
        etag = resp.headers.get("ETag").lower() or resp.headers.get("Content-MD5")
```

and transforming the etag to lowercase.

### Expected

The md5sum of `metadata.json.dvc` is correctly updated.

### Environment information

**Output of `dvc doctor`:**

```console
$ dvc doctor
DVC version: 2.1.0 (conda)
---------------------------------
Platform: Python 3.9.2 on Linux-4.19.0-16-amd64-x86_64-with-glibc2.28
Supports: http, https, ssh
Cache types: hardlink, symlink
Cache directory: ceph on 131.254.135.54:6789,131.254.135.55:6789,131.254.135.56:6789:/volumes/csi/csi-vol-c100e9b9-3eb0-11ea-9c3b-fed11c850da8
Caches: local
Remotes: ssh, local
Workspace directory: ceph on 131.254.135.54:6789,131.254.135.55:6789,131.254.135.56:6789:/volumes/csi/csi-vol-c100e9b9-3eb0-11ea-9c3b-fed11c850da8
Repo: dvc, git
```

**Additional Information (if any):**

**Comment by pmrowla** [comment]
Just to clarify, when I test this locally it looks like `dvc commit` is actually working properly. For weak etags, the uppercase `W/` is [mandatory](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag#directives).
I think problem is that `dvc status` should not be reporting that the checksum for the imported file has changed (and there should be nothing to `dvc commit`)

**Comment by pmrowla** [comment]
Problem is that when loading the `etag` from a dvcfile, we convert all checksum fields to lowercase in

https://github.com/iterative/dvc/blob/6d941e6cb599da71f468fe615243ed66017f34ea/dvc/dvcfile.py#L161
https://github.com/iterative/dvc/blob/6d941e6cb599da71f468fe615243ed66017f34ea/dvc/output/__init__.py#L30-L34

The `stage_text` passed into `validate()` will have the correct capitalized `W/...` etag, but after validate it will be lowercase. This causes a checksum mismatch with what gets written to the dvcfile on import/update - on import the hashinfo for our etag will have the capitalized `W/` and is never passed into `validate()`.

Basically our `validate()`/checksum schema assumes that all of our checksums are case-insensitive and should be converted into lowercase, but this is not actually the case for HTTP etags.

**Comment by kwon-young** [comment]
Thank you for your reply.

> The `stage_text` passed into `validate()` will have the correct capitalized `W/...` etag, but after validate it will be lowercase. This causes a checksum mismatch with what gets written to the dvcfile on import/update - on import the hashinfo for our etag will have the capitalized `W/` and is never passed into `validate()`.
> 
> Basically our `validate()`/checksum schema assumes that all of our checksums are case-insensitive and should be converted into lowercase, but this is not actually the case for HTTP etags.

Ok, so I suppose I can safely ignore what `dvc status` is saying and the problem will fix itself in the next version of dvc ?

**Comment by pmrowla** [comment]
I am not sure when the issue will be fixed, but yes for now you can ignore what `dvc status` says for that imported stage

**Comment by kwon-young** [comment]
So I assume we just need to remove the `Lower` qualification to the `CHECKSUM_SCHEMA` ?
Can I make a pull request ?

**Comment by pmrowla** [comment]
It's not quite that simple, because this issue is specific to HTTP etags. All of the other checksum types that we use in DVC still want the `Lower` in the schema. We will likely need to refactor it to have specific schemas for each possible type instead of using the single general schema that we have right now.

**Comment by kwon-young** [comment]
Well, the thing is that `dvc commit` compares the etag in the dvc file with an unmodified newly downloaded etag.
For me, this means that we should not modify the etag at all.

**Comment by pmrowla** [comment]
> Well, the thing is that `dvc commit` compares the etag in the dvc file with an unmodified newly downloaded etag.
> For me, this means that we should not modify the etag at all.

Right, but this may also mean that we should actually be normalizing (lowercasing everything except the `W/`) the downloaded etag's on import and commit as well, since according to the HTTP spec the only case-sensitive part is the `W/`.

**Comment by pmrowla** [comment]
should also note that this will likely require doing some research into how S3/GS/OSS handle etags as well, as the spec leaves a lot of it up to the web server as far as how to handle etags. So it's possible that we may need to differentiate between them (we currently just use a single general `etag` identifier)

**Comment by kwon-young** [comment]
Maybe we could also add some validation to

https://github.com/iterative/dvc/blob/6d941e6cb599da71f468fe615243ed66017f34ea/dvc/fs/http.py#L149-L153

**Comment by efiop** [comment]
Fixed by #6238
